File: .claude/skills/java-to-kotlin/evals/files/DeviceInfo.java
package jp.developer.bbee.pcassem.model;

import org.springframework.lang.NonNull;

import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.Map;

public record DeviceInfo (String id, String device, String url, String name, String imgurl, String detail,
                          Integer price, Integer rank, int flag1, int flag2,
                          String releasedate, Integer invisible, LocalDateTime createddate, LocalDateTime lastupdate) {

    private static final String DEFAULT_RELEASE_DATE = "20000101";
    private static final LocalDateTime DEFAULT_DATE_TIME = LocalDateTime.of(2000, 1, 1, 0, 0);

    @NonNull
    public static DeviceInfo from(@NonNull Map<String, Object> result) {
        return new DeviceInfo(
                (String) result.get("id"), (String) result.get("device"), (String) result.get("url"),
                (String) result.get("name"), (String) result.get("imgurl"), (String) result.get("detail"),
                toInteger(result.get("price")), toInteger(result.get("rank")),
                toIntOrDefault(result.get("flag1"), 0),
                toIntOrDefault(result.get("flag2"), 0),
                toReleaseDate(result.get("releasedate")), toInteger(result.get("invisible")),
                toLocalDateTime(result.get("createddate"), DEFAULT_DATE_TIME),
                toLocalDateTime(result.get("lastupdate"), DEFAULT_DATE_TIME)
        );
    }

    private static Integer toInteger(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Integer integer) {
            return integer;
        }
        if (value instanceof Number number) {
            return number.intValue();
        }
        return Integer.valueOf(value.toString());
    }

    private static int toIntOrDefault(Object value, int defaultValue) {
        Integer integer = toInteger(value);
        return integer != null ? integer : defaultValue;
    }

    private static String toReleaseDate(Object value) {
        return value != null ? value.toString() : DEFAULT_RELEASE_DATE;
    }

    private static LocalDateTime toLocalDateTime(Object value, LocalDateTime defaultValue) {
        if (value == null) {
            return defaultValue;
        }
        if (value instanceof LocalDateTime localDateTime) {
            return localDateTime;
        }
        if (value instanceof Timestamp timestamp) {
            return timestamp.toLocalDateTime();
        }
        if (value instanceof java.util.Date date) {
            return new Timestamp(date.getTime()).toLocalDateTime();
        }
        return Timestamp.valueOf(value.toString()).toLocalDateTime();
    }
}

File: src/main/java/jp/developer/bbee/pcassem/domain/model/DeviceInfo.kt
package jp.developer.bbee.pcassem.domain.model

import java.sql.Timestamp
import java.time.LocalDateTime

@JvmRecord
data class DeviceInfo(
    val id: String?,
    val device: String?,
    val url: String?,
    val name: String?,
    val imgurl: String?,
    val detail: String?,
    val price: Int?,
    val rank: Int?,
    val flag1: Int,
    val flag2: Int,
    val releasedate: String?,
    val invisible: Int?,
    val createddate: LocalDateTime,
    val lastupdate: LocalDateTime,
) {
    companion object {
        private const val DEFAULT_RELEASE_DATE = "20000101"
        private val DEFAULT_DATE_TIME = LocalDateTime.of(2000, 1, 1, 0, 0)

        @JvmStatic
        fun from(result: Map<String, Any?>): DeviceInfo = DeviceInfo(
            id = result["id"] as? String,
            device = result["device"] as? String,
            url = result["url"] as? String,
            name = result["name"] as? String,
            imgurl = result["imgurl"] as? String,
            detail = result["detail"] as? String,
            price = toInteger(result["price"]),
            rank = toInteger(result["rank"]),
            flag1 = toIntOrDefault(result["flag1"], 0),
            flag2 = toIntOrDefault(result["flag2"], 0),
            releasedate = toReleaseDate(result["releasedate"]),
            invisible = toInteger(result["invisible"]),
            createddate = toLocalDateTime(result["createddate"], DEFAULT_DATE_TIME),
            lastupdate = toLocalDateTime(result["lastupdate"], DEFAULT_DATE_TIME),
        )

        private fun toInteger(value: Any?): Int? = when {
            value == null -> null
            value is Int -> value
            value is Number -> value.toInt()
            else -> value.toString().toInt()
        }

        private fun toIntOrDefault(value: Any?, defaultValue: Int): Int =
            toInteger(value) ?: defaultValue

        private fun toReleaseDate(value: Any?): String =
            value?.toString() ?: DEFAULT_RELEASE_DATE

        private fun toLocalDateTime(value: Any?, defaultValue: LocalDateTime): LocalDateTime = when {
            value == null -> defaultValue
            value is LocalDateTime -> value
            value is Timestamp -> value.toLocalDateTime()
            value is java.util.Date -> Timestamp(value.time).toLocalDateTime()
            else -> Timestamp.valueOf(value.toString()).toLocalDateTime()
        }
    }
}
